### PR TITLE
gh-108751: Add copy.replace() function

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
It creates a modified copy of an object by calling the object's `__replace__()` method.

It is a generalization of dataclasses.replace(), named tuple's _replace() method and replace() methods in various classes, and supports all these stdlib classes.

* Issue: gh-108751

----
:books: Documentation preview :books:: https://cpython-previews--108752.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/copy.rst [hunk 1]
```diff
@@ -89,6 +97,20 @@ with the component as first argument and the memo dictionary as second argument.
 The memo dictionary should be treated as an opaque object.
 
 
+.. index::
+   single: __replace__() (replace protocol)
+
+Function :func:`replace` is more limited than :func:`copy` and :func:`deepcopy`,
+and only supports named tuples, dataclasses, and other classes which
```
Comment (merwok) [inline Doc/library/copy.rst]:
```suggestion
and only supports :term:`named tuples <named tuple>`, :mod:`dataclasses`, and other classes which
```

(I haven’t double checked if `named tuple` or `namedtuple` is the right key)

File: Lib/copy.py [hunk 2]
```diff
@@ -290,3 +290,16 @@ def _reconstruct(x, memo, func, args,
...
+    """Return a new object replacing specified fields with new values.
+
+    This is especially useful for immutable objects, like named tuples or
+    frozen dataclasses.
+    """
+    cls = obj.__class__
+    func = getattr(cls, '__replace__', None)
+    if func is None:
+        raise TypeError(f"replace() does not support {cls.__name__} objects")
+    return func(obj, **changes)
```
Comment (merwok) [inline Lib/copy.py:305]: This is a cool feature for very little code!

**Review by rhettinger — APPROVED** [review]

File: Doc/library/copy.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/library/copy.rst]: Actually, it does not support named tuples in general. It currently only supports named tuples created by `collections.namedtuple()`.

**Comment by vstinner** [comment]
This change introduced a regression: see issue #109052.

With this change, ``test_sys_settrace`` started to crash. I don't know if the bug already existed before and was hidden, or if the ``Objects/codeobject.c`` change really introduced a regression.

File: Doc/whatsnew/3.13.rst [hunk 3]
```diff
@@ -115,6 +115,18 @@ array
   It can be used instead of ``'u'`` type code, which is deprecated.
   (Contributed by Inada Naoki in :gh:`80480`.)
 
+copy
+----
+
+* Add :func:`copy.replace` function which allows to create a modified copy of
+  an object, which is especially usefule for immutable objects.
```
Comment (csm10495) [inline Doc/whatsnew/3.13.rst:122]: usefule -> useful

**Comment by vstinner** [comment]
Would it make sense to add support for dict?

```py
>>> d={'x': 1}
>>> copy.replace(d, x=2)
TypeError: replace() does not support dict objects
```

It would be an alternative to:

```py
>>> d={'x': 1}
>>> dict(d, x=2)
{'x': 2}
>>> {**d, 'x': 2}
{'x': 2}
```

**Comment by vstinner** [comment]
Nice feature.

**Comment by serhiy-storchaka** [comment]
> Would it make sense to add support for dict?

Perhaps no, because it erodes the difference between an object and a dict (as in JavaScript).

And you already have other ways to do this with dicts.

File: Doc/whatsnew/3.13.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Doc/whatsnew/3.13.rst:122]: Thanks. Do you mind to create a PR to fix this typo?

**Pull request merged**